### PR TITLE
Adding telemetry to the dogstatsd client

### PR DESCRIPTION
**Pull request by hush-hush**
Having client side telemetry is very helpful in troubleshooting packet drop between the client and the datadog-agent.

### CODE REVIEW

File: lib/datadog/statsd.rb [hunk 1]
```diff
@@ -560,44 +635,44 @@ def escape_service_check_message(msg)
...
+      end
+    end
 
-        stat = stat.is_a?(String) ? stat.dup : stat.to_s
-        # Replace Ruby module scoping with '.' and reserved chars (: | @) with underscores.
-        stat.gsub!(DOUBLE_COLON, DOT)
-        stat.tr!(':|@'.freeze, UNDERSCORE)
-        full_stat << stat
+    def serialize_metrics(stat, delta, type, opts=EMPTY_OPTIONS, sample_rate=1)
+      full_stat = ''.dup
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply lib/datadog/statsd.rb]: :man_shrugging: I just move this code around, didn't write it. I actually don't need to extract the serialization logic from the `send_stats` method. Putting it back to lower the diff

File: lib/datadog/statsd.rb [hunk 2]
```diff
@@ -20,6 +20,57 @@
 module Datadog
   class Statsd
 
+    class Telemetry
+      attr_accessor :metrics
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply lib/datadog/statsd.rb]: They need to be updated each time we receive a metric, event, service_check. We also need to updated them from the `connection` classes when sending payload over the network.

File: lib/datadog/statsd.rb [hunk 3]
```diff
@@ -560,44 +635,44 @@ def escape_service_check_message(msg)
...
 
-        full_stat << ':'.freeze
-        full_stat << delta.to_s
-        full_stat << PIPE
-        full_stat << type
+      stat = stat.is_a?(String) ? stat.dup : stat.to_s
+      # Replace Ruby module scoping with '.' and reserved chars (: | @) with underscores.
+      stat.gsub!(DOUBLE_COLON, DOT)
+      stat.tr!(':|@'.freeze, UNDERSCORE)
+      full_stat << stat
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply lib/datadog/statsd.rb]: Same here

File: lib/datadog/statsd.rb [hunk 4]
```diff
@@ -33,14 +83,23 @@ class Connection
...
       end
 
       def write(message)
         @logger.debug { "Statsd: #{message}" } if @logger
-        send_message(message)
+        payload = message + @telemetry.flush()
+        send_message(payload)
+
+        @telemetry.reset
+        @telemetry.bytes_sent += payload.length
```
Comment (remeh) [inline lib/datadog/statsd.rb]: Here we reset the values and then we're setting `bytes_sent` and incrementing `packets_sent`. Doesn't it mean that these values will always be equal to what we set them to here? (they're not modified anywhere apart from here and the reset method).

File: lib/datadog/statsd.rb [hunk 5]
```diff
@@ -20,6 +20,56 @@
...
+
+        # using shorthand syntax to reduce the garbage collection
+        return %Q(
+datadog.dogstatsd.client.metrics:#{@metrics}|#{COUNTER_TYPE}|##{@tags}
+datadog.dogstatsd.client.events:#{@events}|#{COUNTER_TYPE}|##{@tags}
+datadog.dogstatsd.client.service_checks:#{@service_checks}|#{COUNTER_TYPE}|##{@tags}
+datadog.dogstatsd.client.bytes_sent:#{@bytes_sent}|#{COUNTER_TYPE}|##{@tags}
+datadog.dogstatsd.client.bytes_dropped:#{@bytes_dropped}|#{COUNTER_TYPE}|##{@tags}
+datadog.dogstatsd.client.packets_sent:#{@packets_sent}|#{COUNTER_TYPE}|##{@tags}
+datadog.dogstatsd.client.packets_dropped:#{@packets_dropped}|#{COUNTER_TYPE}|##{@tags})
```
[the remark this answers is not in the document]
Comment (hush-hush) [reply lib/datadog/statsd.rb]: no need. Line break is dogstatsd metric separator.

File: lib/datadog/statsd.rb (hunk 4 above)
Comment (hush-hush) [reply lib/datadog/statsd.rb]: yes

**Pull request merged**